src/plugins/intel_gpu/src/graph/reduce.cpp

```diff
@@ -95,6 +97,117 @@ layout reduce_inst::calc_output_layout(reduce_node const& node, kernel_impl_para
 }
 
+template<typename ShapeType>
+std::vector<layout> reduce_inst::calc_output_layouts(reduce_node const& /*node*/, kernel_impl_params const& impl_param) {
+    auto desc = impl_param.typed_desc<reduce>();
+
+    auto input0_layout = impl_param.get_input_layout(0);
+
+    // get 'output_shapes' by shape_infer of ngraph
+
+    // input_shapes
+    std::vector<ShapeType> input_shapes = {
+        input0_layout.get<ShapeType>(),
+        ShapeType{0}
+    };
+
+    // output_shapes
+    std::vector<ShapeType> output_shapes = {ShapeType()};
+
+    // const_data
+    auto axes = desc->axes;
+    auto axes_tensor = std::make_shared<ngraph::runtime::HostTensor>(ov::element::i64, ov::Shape{axes.size()}, axes.data());
+    std::map<size_t, std::shared_ptr<ngraph::runtime::HostTensor>> const_data = {{1, axes_tensor}};
+
+    // shape infer by mode
+    auto mode = desc->mode;
+    switch (mode) {
+        case reduce_mode::max:
+        {
+            ov::op::v1::ReduceMax op;
+            shape_infer(&op, input_shapes, output_shapes, const_data);
+            break;
+        }
+        case reduce_mode::min:
+        {
+            ov::op::v1::ReduceMin op;
+            shape_infer(&op, input_shapes, output_shapes, const_data);
+            break;
+        }
+        case reduce_mode::mean:
+        {
+            ov::op::v1::ReduceMean op;
+            shape_infer(&op, input_shapes, output_shapes, const_data);
+            break;
+        }
+        case reduce_mode::prod:
+        {
+            ov::op::v1::ReduceProd op;
+            shape_infer(&op, input_shapes, output_shapes, const_data);
+            break;
+        }
+        case reduce_mode::sum:
+        {
+            ov::op::v1::ReduceSum op;
+            shape_infer(&op, input_shapes, output_shapes, const_data);
+            break;
+        }
+        case reduce_mode::logical_and:
+            {
+            ov::op::v1::ReduceLogicalAnd op;
+            shape_infer(&op, input_shapes, output_shapes, const_data);
+            break;
+        }
+        case reduce_mode::logical_or:
+        {
+            ov::op::v1::ReduceLogicalOr op;
+            shape_infer(&op, input_shapes, output_shapes, const_data);
+            break;
+        }
+        case reduce_mode::l1:
+        {
+            ov::op::v4::ReduceL1 op;
+            shape_infer(&op, input_shapes, output_shapes, const_data);
+            break;
+        }
+        case reduce_mode::l2:
+        {
+            ov::op::v4::ReduceL2 op;
+            shape_infer(&op, input_shapes, output_shapes, const_data);
+            break;
+        }
+        case reduce_mode::sum_square:
+            // not implemented
+        case reduce_mode::log_sum:
+            // not implemented
+        case reduce_mode::log_sum_exp:
+            // not implemented
+        default:
+            CLDNN_ERROR_MESSAGE(
+                desc->id,
+                "Not supported mode");
+    }
```

Comment:
As you mentioned offline, since the reduce's shape infer itself is calculated in its base node, the shape infer results are to be same regardless the mode. So I think you may create only one type of the modes (e.g, prod) add some comments stating the reason why it is used regardless the actual mode used in the node.

Reply:
I'm not 100% sure whether reduce's output shapes are all same for all cases. So I'd like to maintain current switch/case code even though similar codes are repeated.